src/backend/c-lisp.py

```diff
@@ -0,0 +1,197 @@
+        else:
+            return gen_expr(stmt)
+    except Exception as e:
+        print(f"Error in statement: {stmt}")
+        raise e
+
+
+def is_if_stmt(stmt):
+    return stmt[0] == "if"
+
+
+def gen_if_stmt(stmt):
+    rand_suffix = random_label()
+    cond_sym = "tmp_clisp_" + rand_suffix
+    true_lbl = "lbl_true_clisp_" + rand_suffix
+    false_lbl = "lbl_false_clisp_" + rand_suffix
+    out_lbl = "lbl_out_clisp_" + rand_suffix
+
+    cond_instr_list = gen_expr(stmt[1], res_sym=cond_sym)
+    true_instr_list = gen_stmt(stmt[2])
+    if len(stmt) == 3:
+        false_instr_list = []
+    elif len(stmt) == 5 and stmt[3] == "else":
+        false_instr_list = gen_stmt(stmt[4])
+
+    return [
+        *cond_instr_list,
+        ["br", cond_sym, true_lbl, false_lbl],
+        ["label", true_lbl],
+        *true_instr_list,
+        ["jmp", out_lbl],
+        ["label", false_lbl],
+        *false_instr_list,
+        ["label", out_lbl],
+    ]
+
+
+def is_decl_stmt(stmt):
+    return stmt[0] == "declare"
+
+
+def gen_decl_stmt(stmt):
+    if not (len(stmt) == 2 and len(stmt[1]) == 2):
+        raise CodegenError(f"bad declare statement: {stmt}")
+
+    symbol_types[stmt[1][0]] = stmt[1][1]
+    return []
+
+
+def is_ret_stmt(stmt):
+    return stmt[0] == "ret"
+
+
+def gen_ret_stmt(stmt):
+    if len(stmt) == 1:
+        return ["ret"]
+    elif len(stmt) == 2:
+        res_sym = random_label("tmp_clisp")
+        instr_list = gen_expr(stmt[1], res_sym=res_sym)
+        instr_list.append(["ret", res_sym])
+        return instr_list
+    else:
+        raise CodegenError(
+            f"Return statement can contain only 1 optional return value: {stmt}"
+        )
+
+
+def is_compound_stmt(stmt):
+    return isinstance(stmt, list) and isinstance(stmt[0], list)
+
+
+def gen_compound_stmt(stmt):
+    instr_list = []
+    for s in stmt:
+        instr_list += gen_stmt(s)
+    return instr_list
+
+
+def is_set_expr(expr):
+    return expr[0] == "set"
+
+
+def gen_set_expr(expr, res_sym):
+    name = expr[1]
+    if not name in symbol_types:
+        raise CodegenError(f"Cannot set undeclared variable: {name}")
+
+    instr_list = gen_expr(expr[2], res_sym=res_sym)
+    instr_list.append(["set", [name, symbol_types[name]], ["id", res_sym]])
+    return instr_list
+
+
+def get_literal_type(expr):
+    if isinstance(expr, bool):
+        return "bool"
+    elif isinstance(expr, int):
+        return "int"
+    else:
+        return False
+
+
+def gen_literal_expr(expr, res_sym):
+    return [["set", [res_sym, get_literal_type(expr)], ["const", expr]]]
+
+
+def is_call_expr(expr):
+    return expr[0] == "call"
+
+
+def gen_call_expr(expr, res_sym):
+    instr_list = []
+    arg_syms = []
+    for arg in expr[2:]:
+        arg_sym = random_label("tmp_clisp")
+        arg_syms.append(arg_sym)
+        instr_list += gen_expr(arg, res_sym=arg_sym)
+    instr_list.append(["set", [res_sym, "int"], ["call", expr[1], *arg_syms]])
```

Comment:
why are you doing "int" here?

Reply:
Looks like we do need type inference. Declare does look useless, sorry.

Reply:
This is because the implementation for function definitions is incomplete. Eventually, we will store the return type of all functions in a table; and we will look up that table to fill in what is hard-coded as "int" here.

Reply:
something like
`(define ((fn bool)))` -> thus we know that `fn` returns `bool`
`(set res (call fn))` -> we know that `fn` returns `bool`, thus `res` must be a `bool`

Reply:
> Looks like we do need type inference.

I think we can get away without it
* function return types are fixed
* If we keep declare and track variabe types, then variable types are also fixed.
* a lot binary expressions have fixed result types (e.g. comparison is boolean). If we do away with C's implicit casting, the rules for inferring types of arithmetic expressions become dead simple (essentially, result type is same as input type)

Reply:
Storing function types is exactly what I mean by type inference.

Reply:
Changed

Reply:
Function return and parameter types are now tracked.